### PR TITLE
[Merged by Bors] - added running hash for validated layers

### PR DESCRIPTION
**Pull request by antonlerner**
To support new sync features we need a running hash of layers

### CODE REVIEW

File: mesh/mesh.go [hunk 1]
```diff
@@ -304,10 +304,25 @@ func (msh *Mesh) pushLayersToState(oldPbase types.LayerID, newPbase types.LayerI
...
+
+func (msh *Mesh) persistLayerHashes(l *types.Layer) {
+	hash := msh.calcLayerHash(l)
+	msh.persistLayerHash(l.Index(), hash)
+	prevHash := types.Hash32{}
+	var err error
+	if l.Index() > types.GetEffectiveGenesis() {
+		prevHash, err = msh.getRunningLayerHash(l.Index() - 1)
+		if err != nil {
+			msh.Log.Error("cannot get running layer hash for layer %v", l.Index()-1)
```
Comment (y0sher) [inline mesh/mesh.go:321]: this means later prevHash will be nil..

File: mesh/mesh.go [hunk 2]
```diff
@@ -420,21 +435,55 @@ func (msh *Mesh) logStateRoot(layerID types.LayerID) {
...
 	if err := msh.general.Put(constLAYERHASH, msh.layerHash); err != nil {
+		msh.With().Error("failed to persist last layer hash", log.Err(err), msh.ProcessedLayer(),
+			log.String("layer_hash", util.Bytes2Hex(msh.layerHash)))
+	}
+}
+
+func (msh *Mesh) persistLayerHash(layerID types.LayerID, hash types.Hash32) {
+	if err := msh.general.Put(msh.getLayerHashKey(layerID), hash.Bytes()); err != nil {
 		msh.With().Error("failed to persist layer hash", log.Err(err), msh.ProcessedLayer(),
 			log.String("layer_hash", util.Bytes2Hex(msh.layerHash)))
```
Comment (y0sher) [inline mesh/mesh.go]: we should log the persisted layer hash here instead of the latest..

**Review by y0sher — APPROVED** [review]

**Comment by antonlerner** [comment]
bors try

**Comment by antonlerner** [comment]
bors merge

**Comment by antonlerner** [comment]
bors merge

**Pull request closed without merging**